test/metabase/lib/query_test.cljc

```diff
@@ -506,3 +508,26 @@
                                     "1969-10-12"
                                     "1971-10-12"]]}]}
               (lib/query mp query))))))
+
+#?(:clj
+   (deftest ^:synchronized cache-test
+     (let [query      (lib/query meta/metadata-provider (meta/table-metadata :orders))
+           viz-cols   lib.metadata.calculation/visible-columns-method
+           calls      (atom 0)
+           exp-fields (into #{} cat
+                            [(map #(meta/id :orders %)   (meta/fields :orders))
+                             (map #(meta/id :people %)   (meta/fields :people))
+                             (map #(meta/id :products %) (meta/fields :products))])]
+       (is (=? {} (-> query meta :lib/__cache)))
+       (with-redefs [lib.metadata.calculation/visible-columns-method
+                     (fn [query stage-number x options]
+                       (when (= x (lib.util/query-stage query stage-number))
+                         (swap! calls inc))
+                       (viz-cols query stage-number x options))]
+         (is (=? 0 @calls))
+         (is (=? exp-fields
+                 (into #{} (map :id) (lib/visible-columns query))))
+         (is (=? 1 @calls))
+         (is (=? exp-fields
+                 (into #{} (map :id) (lib/visible-columns query))))
+         (is (=? 1 @calls))))))
```

Comment:
I think these assertions should just use `=`. That's shorter, and nobody would have to think how `=?` is defined for sets.

Reply:
Done.